File: M/2144-minimum-cost-of-buying-candies-with-discount/main.go
package main

import "sort"

func minimumCost(cost []int) int {
	sort.Ints(cost)

	sum := 0
	for i := len(cost) - 1; i >= 0; i-- {
		if (len(cost)-1-i)%3 != 2 {
			sum += cost[i]
		}
	}
	return sum
}

func minimumCost2(cost []int) int {
	sort.Ints(cost)
	buf := make([]int, len(cost))

	sum := 0
	for i := len(cost) - 1; i >= 0; i-- {
		if buf[i] == 1 {
			continue
		}
		sum += cost[i]
		buf[i] = 1
		idx := -1
		for j := i - 1; j >= 0; j-- {
			if buf[j] == 0 {
				idx = j
				break
			}
		}
		if idx > -1 {
			sum += cost[idx]
			buf[idx] = 1
			gidx := -1
			for j := idx - 1; j >= 0; j-- {
				if buf[j] == 0 {
					gidx = j
					break
				}
			}
			if gidx > -1 {
				buf[gidx] = 1
			}
		}
	}
	return sum
}

File: M/2144-minimum-cost-of-buying-candies-with-discount/main.rs
impl Solution {
  pub fn minimum_cost(mut cost: Vec<i32>) -> i32 {
    let mut idx: i32 = cost.len() as i32 - 1;
    cost.sort_unstable();
    let mut sum: i32 = 0;
    while idx >= 0 {
      sum += if idx >= 0 { cost[idx as usize] } else { 0 } + if idx >= 1 { cost[idx as usize - 1] } else { 0 };
      idx -= 3;
    }
    sum
  }
}
